pkg/alertmanager/alertmanager.go

```diff
@@ -575,3 +607,117 @@ func (t *tenantRateLimits) RateLimit() rate.Limit {
 func (t *tenantRateLimits) Burst() int {
 	return t.limits.NotificationBurstSize(t.tenant, t.integration)
 }
+
+var (
+	errTooManyAlerts = "too many alerts, limit: %d"
+	errAlertsTooBig  = "alerts too big, total size limit: %d bytes"
+)
+
+const (
+	insertFailureTooManyAlerts = "too_many_alerts"
+	insertFailureAlertsTooBig  = "alerts_too_big"
+)
+
+type alertsLimiter struct {
+	tenant string
+	limits Limits
+
+	failureCounter *prometheus.CounterVec
+
+	mx        sync.Mutex
+	sizes     map[model.Fingerprint]int
+	count     int
+	totalSize int
+}
+
+func newAlertsLimiter(tenant string, limits Limits, failureCounter *prometheus.CounterVec) *alertsLimiter {
+	return &alertsLimiter{
+		tenant:         tenant,
+		limits:         limits,
+		sizes:          map[model.Fingerprint]int{},
+		failureCounter: failureCounter,
+	}
+}
+
+func (a *alertsLimiter) PreStore(alert *types.Alert, existing bool) error {
+	if alert == nil {
+		return nil
+	}
+
+	countLimit := a.limits.AlertmanagerMaxAlertsCount(a.tenant)
+	sizeLimit := a.limits.AlertmanagerMaxAlertsSizeBytes(a.tenant)
+
+	newSize := alertSize(alert.Alert)
+
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	if !existing {
```

Comment:
An alert is considered unique based on the hash of its labels (fingerprint), technically if an alert comes in with a size that's OK, then its annotations change and its size is no longer acceptable we would bypass it because we would consider its size to be OK because it was already there.

I believe we need to check that if it does exist, its new size would not tip us over the limit. Unless you want the semantic to be: **If we had the alert before, we'll always accept it regardless of whenever it tips us over the limit**

If this is the case, can we add a comment?

Reply:
> If we had the alert before, we'll always accept it regardless of whenever it tips us over the limit

That was indeed my intention, mostly to make sure that we don't accidentally reject "resolved" alerts. But if the size has changed, it may make sense to reject it if it's over limit. I don't have strong opinion on this.

Reply:
Also ... when new alert arrives and "merges" with existing one, it cannot update labels (fingerprint would change) or annotations (right now, at least).

See the merging code: https://github.com/cortexproject/cortex/blob/5e496f428302a195a68cc287d97641561dbe0afd/vendor/github.com/prometheus/alertmanager/types/types.go#L366

Reply:
> or annotations (right now, at least)

I see that is true for the labels, which is expected but is it also for the annotations? I can't seem to see it in the code - am I missing something?

Reply:
> I see that is true for the labels, which is expected but is it also for the annotations? I can't seem to see it in the code - am I missing something?

No, you’re right. I misread the code I linked… line 372 is copying everything over, including annotations.

Reply:
Makes sense. In this case, I think rejection would be the best course of action - don't you think?

The alert that existed before should have a short enough `endsAt` that it would resolve itself eventually so there's little to no risk of just leaving the existing one to resolve by itself. On the contrary, accepting the alert that might tip us over the limit risks our availability.

Reply:
I think you're right, and I will make this change before merging.

Reply:
I've changed the logic such that existing alert that grows and doesn't fit the limit anymore if rejected.